include/fastrtps/utils/collections/ResourceLimitedVector.hpp

```diff
@@ -0,0 +1,337 @@
+    using iterator = typename collection_type::iterator;
+    using const_iterator = typename collection_type::const_iterator;
+    using reverse_iterator = typename collection_type::reverse_iterator;
+    using const_reverse_iterator = typename collection_type::reverse_iterator;
+
+    /**
+     * Construct a ResourceLimitedVector.
+     *
+     * This constructor receives a \ref ResourceLimitsConfig to setup the allocation behaviour regarding the 
+     * number of elements in the collection.
+     *
+     * The cfg parameter indicates the initial number to be reserved, the maximum number of items allowed,
+     * and the capacity increment value.
+     *
+     * @param cfg     Resource limits configuration to use.
+     * @param alloc   Allocator object. Forwarded to collection constructor.
+     */
+    ResourceLimitedVector(
+            ResourceLimitedContainerConfig cfg = {}, 
+            const allocator_type& alloc = allocator_type())
+        : configuration_(cfg)
+        , collection_(alloc)
+    {
+        collection_.reserve(cfg.initial);
+    }
+
+    /**
+     * Add element at the end.
+     *
+     * Adds a new element at the end of the vector, after its current last element. 
+     * The content of val is copied to the new element.
+     *
+     * @param val   Value to be copied to the new element.
+     *
+     * @return pointer to the new element, nullptr if resource limit is reached.
+     */
+    pointer push_back(const value_type& val)
+    {
+        return emplace_back(val);
+    }
+
+    /**
+     * Add element at the end.
+     *
+     * Adds a new element at the end of the vector, after its current last element. 
+     * The content of val is moved to the new element.
+     *
+     * @param val   Value to be moved to the new element.
+     *
+     * @return pointer to the new element, nullptr if resource limit is reached.
+     */
+    pointer push_back(value_type&& val)
+    {
+        return emplace_back(std::move(val));
+    }
+
+    /**
+     * Construct and insert element at the end.
+     *
+     * Inserts a new element at the end of the vector, right after its current last element. 
+     * This new element is constructed in place using args as the arguments for its constructor.
+     *
+     * @param args   Arguments forwarded to construct the new element.
+     *
+     * @return pointer to the new element, nullptr if resource limit is reached.
+     */
+    template<typename ... Args>
+    pointer emplace_back(Args&& ... args)
+    {
+        if (!ensure_capacity())
+        {
+            // Indicate error by returning null pointer
+            return nullptr;
+        }
+
+        // Keep record of insertion position
+        size_type previous_size = collection_.size();
+
+        // Construct new element at the end of the collection
+        collection_.emplace_back(args...);
+
+        // Return pointer to newly created element
+        return &collection_[previous_size];
```

Comment:
In this case, I think it would be better to `return &collection_.back();`, and forget about the `previous_size`.